File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/java/IdeClasspathResolver.java
package org.jetbrains.bsp.bazel.server.sync.languages.java;

import io.vavr.collection.Seq;
import io.vavr.collection.Set;
import io.vavr.control.Option;
import java.net.URI;
import java.util.regex.Pattern;

public class IdeClasspathResolver {
  private static final Pattern JAR_PATTERN = Pattern.compile("-[hi]jar\\.jar$");

  private final Set<String> runtimeJars;
  private final Set<String> runtimeMavenJarSuffixes;
  private final Seq<String> compileJars;

  public IdeClasspathResolver(Seq<URI> runtimeClasspath, Seq<URI> compileClasspath) {
    this.runtimeJars = runtimeClasspath.map(URI::toString).toSet();
    this.runtimeMavenJarSuffixes = runtimeJars.flatMap(this::toMavenSuffix);
    this.compileJars = compileClasspath.map(URI::toString);
  }

  public Seq<URI> resolve() {
    return compileJars.iterator().map(this::findRuntimeEquivalent).map(URI::create).toArray();
  }

  private String findRuntimeEquivalent(String compileJar) {
    var runtimeJar = JAR_PATTERN.matcher(compileJar).replaceAll(".jar");
    if (runtimeJars.contains(runtimeJar)) {
      return runtimeJar;
    }

    var headerSuffix = toMavenSuffix(compileJar);
    var mavenJarSuffix = headerSuffix.map(s -> s.replaceAll("/header_([^/]+)\\.jar$", "/$1.jar"));
    if (mavenJarSuffix.exists(runtimeMavenJarSuffixes::contains)) {
      return runtimeJars.find(jar -> jar.endsWith(mavenJarSuffix.get())).get();
    }
    return compileJar;
  }

  private Option<String> toMavenSuffix(String uri) {
    var indicator = "/maven2/";
    int index = uri.lastIndexOf(indicator);
    if (index < 0) {
      return Option.none();
    } else {
      return Option.some(uri.substring(index + indicator.length()));
    }
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/java/IdeClasspathResolver.kt
package org.jetbrains.bsp.bazel.server.sync.languages.java

import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver
import org.jetbrains.bsp.bazel.server.sync.model.Label
import java.net.URI

class IdeClasspathResolver(
  private val label: Label,
  private val bazelPathsResolver: BazelPathsResolver,
  runtimeClasspath: Sequence<URI>,
  compileClasspath: Sequence<URI>,
) {
  private val runtimeJars: Set<String>
  private val runtimeMavenJarSuffixes: Set<String>
  private val compileJars: Sequence<String>

  init {
    runtimeJars = runtimeClasspath.map { it.toString() }.toSet()
    runtimeMavenJarSuffixes = runtimeJars.mapNotNull(::toMavenSuffix).toSet()
    compileJars = compileClasspath.map { it.toString() }
  }

  fun resolve(): Sequence<URI> =
    compileJars
      .map(::findRuntimeEquivalent)
      .filterNot { isItJarOfTheCurrentTarget(it) }
      .map(URI::create)

  private fun findRuntimeEquivalent(compileJar: String): String {
    val runtimeJar = compileJar.replace(JAR_PATTERN, ".jar")
    if (runtimeJars.contains(runtimeJar)) {
      return runtimeJar
    }
    val headerSuffix = toMavenSuffix(compileJar)
    val mavenJarSuffix = headerSuffix?.let { s: String ->
      s.replace(
        "/header_([^/]+)\\.jar$".toRegex(),
        "/$1.jar"
      )
    }
    return mavenJarSuffix?.takeIf(runtimeMavenJarSuffixes::contains)?.let { suffix ->
      runtimeJars.find { jar: String -> jar.endsWith(suffix) }
    } ?: compileJar
  }

  private fun toMavenSuffix(uri: String): String? {
    val indicator = "/maven2/"
    val index = uri.lastIndexOf(indicator)
    return if (index < 0) {
      null
    } else {
      uri.substring(index + indicator.length)
    }
  }

  private fun isItJarOfTheCurrentTarget(jar: String): Boolean {
    val targetPath = bazelPathsResolver.extractRelativePath(label.value)
    val targetJar = "$targetPath/${label.targetName()}.jar"

    return jar.endsWith(targetJar)
  }

  companion object {
    private val JAR_PATTERN = ("((-[hi]jar)|(\\.abi))\\.jar\$").toRegex()
  }
}
